lib/utilities/pipeline.js

```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+var Promise = require('rsvp').Promise;
+var _       = require('lodash-node');
+
+var chalk = require('chalk');
+var blue  = chalk.blue;
+var red   = chalk.red;
+var grey  = chalk.grey;
+
+function Pipeline(hookNames, options) {
+  hookNames = hookNames || [];
+  options = options || {};
+
+  this._ui = options.ui;
+
+  this._pipelineHooks = hookNames.reduce(function(pipelineHooks, hookName) {
+    pipelineHooks[hookName] = [];
+
+    return pipelineHooks;
+  }, {});
+}
+
+Pipeline.prototype.register = function(hookName, fn) {
+  var pipelineHooks = this._pipelineHooks;
+
+  if (pipelineHooks[hookName]) {
+    pipelineHooks[hookName].push(fn);
+  }
+};
+
+Pipeline.prototype.execute = function(context) {
```

Comment:
This all gets pretty gnarly.  I have been trying to make it uncomplicated but not sure how to.  Still, it's a work in progress so it will change.  Happy to take comments, suggestions, advice.